File: tools-py/tasks/indexes.py
import os, sys
from tools import common
from pathlib import Path
from datetime import datetime

export_regexes = [
    common.regex.EXPORT_DECL,
    common.regex.EXPORT_LIST
]

def _generate_indexes_internal(directory: Path):
    if not directory.exists():
        print("Directory " + directory.__str__() + " does not exist")
        sys.exit(1)

    print("Generating index.js file for directory " + directory.__str__())
    exports = []
    order_entries = common._read_order_file(directory, 'js')
    file_list = [f for f in directory.iterdir() if f.suffix in [".js", ".mjs"]]

    ordered_file_list = []

    for order_file in order_entries:
        for file in file_list:
            if file.name == order_file:
                ordered_file_list.append(file)
                file_list.remove(file)
                break
    for file in file_list:
        ordered_file_list.append(file)

    for file in ordered_file_list:
        if file.name == "index.js":
            continue
        with open(file, 'r', encoding='utf-8') as f:
            content = f.read()
        base_name = file.stem
        file_exports = []
        for regex in export_regexes:
            for match in regex.finditer(content):
                raw_names = match.group(1).split(',')
                for name in raw_names:
                    #parts = re.split(r"\s+as\s+", name.strip())
                    parts = common.regex.EXPORT_AS.split(name.strip())
                    export_name = parts[1] if len(parts) > 1 else parts[0]
                    file_exports.append(export_name)
        joined_file_exports = ", ".join(file_exports)
        exports.append(f"export {{{joined_file_exports}}} from './{base_name}.js';")
    now = datetime.now().astimezone()
    offset = now.utcoffset()
    offset_hours = offset.total_seconds() / 3600
    index_file = directory / "index.js"
    created_file = not index_file.exists()

    header = (
            "// File generated automatically.\n"
            "// Last Updated: " + now.strftime(
        f"%d/%m/%Y %H:%M:%S.{now.microsecond // 1000:03d} UTC{offset_hours:+.0f}") + "\n\n"
    )
    body = "console.log(`Loaded: ${import.meta.url}`);\n\n" + "\n".join(exports) + "\n"
    new_content = header + body
    if exports.__len__() > 0:
        if not created_file:
            existing = index_file.read_text(encoding="utf-8")
            # Normalize to line lists and drop the first two lines for comparison
            existing_lines = existing.splitlines()
            new_lines = new_content.splitlines()
            existing_body = "\n".join(existing_lines[2:])  # ignore first 2 comment lines
            new_body = "\n".join(new_lines[2:])  # ignore first 2 comment lines
            if existing_body != new_body:
                index_file.write_text(new_content, encoding="utf-8")
        if created_file:
            print(f"adding {index_file} to git")
            index_file.write_text(new_content, encoding="utf-8")
            os.system(f"git add {index_file}")

def generate_indexes(file: Path, dira: Path):
    if not(file is None):
        _generate_indexes_internal(Path(file).resolve().parent)
    elif not(dira is None):
        _generate_indexes_internal(Path(dira))
    else:
        root = common.project_root() / "unofficial-FVTT-ose" / "scripts"
        for dir_path, dir_names, file_names in os.walk(root):
            _generate_indexes_internal(Path(dir_path).resolve())
    sys.exit(0)

File: tools/tasks/indexes.js
import fs from 'node:fs';
import path from 'node:path';
import * as common from '../common.js';
import {collectExports} from '../lib/jsExports.js';

console.log(`Loaded: ${import.meta.url}`);

/**
 * @param {string} directory - absolute path
 * @returns {'created'|'changed'|'unchanged'|'empty'}
 */
function generateIndexesInternal(directory) {
    if (!fs.existsSync(directory)) {
        console.log(`Directory ${directory} does not exist`);
        process.exit(1);
    }

    // The tooling is not a module tree. Generating a barrel here would overwrite
    // the CLI entry point with generated exports and break every invocation.
    if (common.isInsideToolsRoot(directory)) {
        common.fail(`Refusing to generate an index.js inside the tooling folder: ${directory}`);
    }

    console.log(`Generating index.js file for directory ${directory}`);

    const orderEntries = common.readOrderFile(directory, 'js');
    let fileList = fs.readdirSync(directory, {withFileTypes: true})
        .filter(f => f.isFile() && ['.js', '.mjs'].includes(path.extname(f.name)))
        .map(f => path.join(directory, f.name));

    const orderedFileList = [];
    for (const orderFile of orderEntries) {
        const idx = fileList.findIndex(f => path.basename(f) === orderFile);
        if (idx !== -1) {
            orderedFileList.push(fileList[idx]);
            fileList.splice(idx, 1);
        }
    }
    orderedFileList.push(...fileList);

    const exports = [];
    for (const file of orderedFileList) {
        if (path.basename(file) === 'index.js') continue;
        const baseName = path.basename(file, path.extname(file));
        const fileExports = collectExports(file);
        if (fileExports === null) continue;
        const joinedFileExports = fileExports.join(', ');
        exports.push(`export {${joinedFileExports}} from './${baseName}.js';`);
    }

    const now = common.nowTmz();
    const indexFile = path.join(directory, 'index.js');
    const createdFile = !fs.existsSync(indexFile);

    const header = `// File generated automatically.\n// Last Updated: ${common.formatHeaderTimestamp(now)}\n\n`;
    const body = 'console.log(`Loaded: ${import.meta.url}`);\n\n' + exports.join('\n') + '\n';
    const newContent = header + body;

    if (exports.length === 0) return 'empty';

    if (createdFile) {
        fs.writeFileSync(indexFile, newContent, 'utf-8');
        console.log(`[CREATED] ${indexFile} (staged with git add)`);
        common.run(['git', 'add', indexFile]);
        return 'created';
    }

    // Compare below the two-line header, so a timestamp alone is not a change
    const existingBody = fs.readFileSync(indexFile, 'utf-8').split('\n').slice(2).join('\n');
    const newBody = newContent.split('\n').slice(2).join('\n');
    if (existingBody === newBody) return 'unchanged';

    fs.writeFileSync(indexFile, newContent, 'utf-8');
    console.log(`[CHANGED] ${indexFile}`);
    return 'changed';
}

/**
 * @param {string|null} file - filename of the changed file; index.js of that folder is generated
 * @param {string|null} dir - directory to generate index.js for
 */
export async function generateIndexes(file, dir) {
    const tally = {created: 0, changed: 0, unchanged: 0, empty: 0};
    const record = (result) => {
        tally[result] += 1;
    };

    if (file != null) {
        record(generateIndexesInternal(path.dirname(path.resolve(file))));
    } else if (dir != null) {
        record(generateIndexesInternal(path.resolve(dir)));
    } else {
        const root = path.join(common.projectRoot(), 'unofficial-FVTT-ose', 'scripts');
        const stack = [root];
        while (stack.length) {
            const dirPath = stack.shift();
            record(generateIndexesInternal(dirPath));
            for (const entry of fs.readdirSync(dirPath, {withFileTypes: true})) {
                if (entry.isDirectory()) stack.push(path.join(dirPath, entry.name));
            }
        }
    }

    const touched = tally.created + tally.changed;
    console.log(`indexes: ${touched} file(s) updated`
        + ` (${tally.created} created, ${tally.changed} changed,`
        + ` ${tally.unchanged} unchanged, ${tally.empty} with no exports).`);

    process.exit(0);
}
